### ISSUE TITLE
didReceive Remote Message never called

### ISSUE THREAD

**Issue by Chris59160** [firebase/firebase-ios-sdk#693]
func application(_ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable : Any], fetchCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Swift.Void)

func messaging(_ messaging: Messaging, didReceive remoteMessage: MessagingRemoteMessage)
### [READ] Step 1: Are you in the right place?

Yes

### [REQUIRED] Step 2: Describe your environment

  * Xcode version: 9.2
  * Firebase 4.8.1
  * FirebaseAnalytics 4.0.7
  * FirebaseCore 4.0.14
  * FirebaseInstanceID 2.0.8
  * FirebaseMessaging 2.0.8

### [REQUIRED] Step 3: Describe the problem

I have 2 problems.

First one is that When I sent a message to a device, whether with a payload for notification or not, it always is receive in

```
func application(_ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable : Any], fetchCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Swift.Void)
```

The only way I managed to get 

```
func messaging(_ messaging: Messaging, didReceive remoteMessage: MessagingRemoteMessage)
```
 
called, was when I run the app on a simulator, otherwise on a real device, it's always "didReceiveRemoteNotification".

Maybe the second problem is related some how, I have used the Firebase Message console to send a notification to my device. I can receive it if the app is in foreground. In the background, I will get it only when I maximise the app....and if the app isn't running, I have no notifications showing up on my device, and I don't know why.

I think some how the device is registered for data only, but not for really notification which would explain why I can't get any notification when app is killed.

If I set shouldEstablishDirectChannel to false, I can't receive any thing.

I think some how, my configuration is wrong, but can't work out what's wrong.

I have attached my AppDelegate.

#### Steps to reproduce:

This is how I set things up:

```
        FirebaseApp.configure()
    
        Messaging.messaging().delegate = self
        Messaging.messaging().shouldEstablishDirectChannel = true
        Messaging.messaging().subscribe(toTopic: "topicEventSlotUpdated")
        
        self.registerForNotification()

func registerForNotification() -> Void
    {
        if #available(iOS 10.0, *)
        {
            let notificationCenter:UNUserNotificationCenter = UNUserNotificationCenter.current()
            notificationCenter.delegate = self
            notificationCenter.requestAuthorization(options: [.alert, .sound, .badge]) {
                (granted, error) in
                // Enable or disable features based on authorization.
                print(error as Any)
                if granted
                {
                    print("Register for APS")
                    UIApplication.shared.registerForRemoteNotifications()
                }
                else
                {
                    print("Failed to register for APS")
                }
            }
        }
        else
        {
            // Fallback on earlier versions
            let setting:UIUserNotificationSettings = UIUserNotificationSettings(types: [.alert, .badge, .sound], categories: nil)
            UIApplication.shared.registerUserNotificationSettings(setting)
            UIApplication.shared.registerForRemoteNotifications()
        }
    }
```
I am implementing MessagingDelegate in my AppDelegate

```
func messaging(_ messaging: Messaging, didReceiveRegistrationToken fcmToken: String)
    {
        print("Firebase registration token: \(fcmToken)")
        // TODO: If necessary send token to application server.
        // Note: This callback is fired at each app startup and whenever a new token is generated.
        AccountApi.setNotificationToken(notifToken: fcmToken)
    }
    
    func application(received remoteMessage: MessagingRemoteMessage)
    {
        print("applicationReceivedRemoteMessage")
        print(remoteMessage)
        print(remoteMessage.appData)
    }
    
    func messaging(_ messaging: Messaging, didReceive remoteMessage: MessagingRemoteMessage)
    {
        // Convert to pretty-print JSON, just to show the message for testing
        guard let data = try? JSONSerialization.data(withJSONObject: remoteMessage.appData, options: .prettyPrinted),
            let prettyPrinted = String(data: data, encoding: .utf8) else {
                return
        }
        print("Received direct channel message:\n\(prettyPrinted)")
    }
```
#### Relevant Code:

Interesting to note when I received the message in the "func application(_ application: UIApplication, didReceiveRemoteNotification", I don't have the notification details.

Keep in mind, in my code, I always send things twice, so first time without any notification payload, to send only some data, then the second message includes the Notification payload, to make sure the device will get the data.

On Android, I always receive the message twice if I am in the foreground, but on iOS, I receive it once only...

[AppDelegate.swift.zip](https://github.com/firebase/firebase-ios-sdk/files/1657150/AppDelegate.swift.zip)

**Comment by Chris59160** [comment]
Digging in Firebase code, I found out:

`/// This method is called on iOS 10 devices to handle data messages received via FCM through its
/// direct channel (not via APNS). For iOS 9 and below, the FCM data message is delivered via the
/// UIApplicationDelegate's -application:didReceiveRemoteNotification: method.
- (void)messaging:(nonnull FIRMessaging *)messaging
    didReceiveMessage:(nonnull FIRMessagingRemoteMessage *)remoteMessage
    NS_SWIFT_NAME(messaging(_:didReceive:))
    __IOS_AVAILABLE(10.0);

/// The callback to handle data message received via FCM for devices running iOS 10 or above.
- (void)applicationReceivedRemoteMessage:(nonnull FIRMessagingRemoteMessage *)remoteMessage
    NS_SWIFT_NAME(application(received:))
    __deprecated_msg("Use FIRMessagingDelegate’s -messaging:didReceiveMessage:");`

So that's why I receive every thing on my device running iOS 9.2.3

My apology for that, I should have dug up the code better...but it should be mentioned in the doc really.

**Comment by morganchen12** [comment]
Not being able to receive messages when disabling direct channel indicates there's something wrong with your configuration, so you may still have messaging issues after implementing those two methods.

Are you able to send messages directly through APNs without FCM?

**Comment by Chris59160** [comment]
I am using another engine in parallel called Ably, and the notifications always went through with that engine.
The issue I have now is that I don't receive notification in background or when app is not running. ( i.e I don't have the usual Notification banner )

I am running in debug, and my APNS certificate for debug is correct and uploaded on FCM. I use the same with this other engine and the Notifications do work with that engine.

I am basically trying to remove this Ably engine, to use FCM completely for real-time messages and Notifications :-)

**Comment by morganchen12** [comment]
What notification payloads are you trying to send through FCM? Do you get any errors in the FCM console when trying to send those messages?

**Comment by Chris59160** [comment]
Ok, I have sorted it out.

For some reasons, I was trying to send the Notification payload wrongly. I added aps key, etc..myself when FCM is managing that themselves, with Notification payload being identical between Android and iOS for the end user such as myself.

I got confused by all the suggestion from people online on why it couldn't work.

Now every thing is fine...and I am happy :-)

Sorry for posting this, it was useless in the end....

Thank you for your time!
